### PR TITLE
meson: add preliminary support for cross compilation

### PR DESCRIPTION
**Pull request by pbui**
This should address (or begin to address) #7656.

I've tested this with armv6, armv6l-musl, armv7l for the ncmpc and fuse-sshfs packages.

### CODE REVIEW

File: common/build-style/meson.sh [hunk 1]
```diff
@@ -4,6 +4,55 @@
 do_configure() {
 	: ${meson_cmd:=meson}
 	: ${meson_builddir:=build}
+	: ${meson_crossfile:=meson.cross}
+
+	if [ "$CROSS_BUILD" ]; then
+		_MESON_TARGET_ENDIAN=little
+		_MESON_TARGET_CPU=${XBPS_TARGET_MACHINE}
+		case "$XBPS_TARGET_MACHINE" in
+			mips)
```
Comment (Vaelatern) [inline common/build-style/meson.sh]: `mips*` perhaps?

File: common/build-style/meson.sh [hunk 2]
```diff
@@ -4,6 +4,55 @@
 do_configure() {
 	: ${meson_cmd:=meson}
 	: ${meson_builddir:=build}
+	: ${meson_crossfile:=meson.cross}
```
Comment (Vaelatern) [inline common/build-style/meson.sh]: can we make this perhaps `xbps_meson.cross` or something like that? Something less generic.

File: common/build-style/meson.sh [hunk 3]
```diff
@@ -4,6 +4,55 @@
...
+				_MESON_TARGET_ENDIAN=big
+				;;
+
+		esac
+
+		# Record cross-compiling information in cross file.
+		# CFLAGS and LDFLAGS must be set as c_args and c_link_args.
+		cat > ${meson_crossfile} <<EOF
+[binaries]
+c = '${XBPS_CROSS_TRIPLET}-gcc'
```
Comment (Vaelatern) [inline common/build-style/meson.sh]: what's wrong with `${CC}`?

File: common/build-style/meson.sh [hunk 4]
```diff
@@ -4,6 +4,55 @@
...
+cpp = '${XBPS_CROSS_TRIPLET}-g++'
+ar = '${XBPS_CROSS_TRIPLET}-ar'
+ld = '${XBPS_CROSS_TRIPLET}-ld'
```
Comment (Vaelatern) [inline common/build-style/meson.sh]: Maybe all this can just use the XBPS vars instead of reconstructing?

File: common/build-style/meson.sh (hunk 1 above)
Comment (Johnnynator) [reply common/build-style/meson.sh]: Isn't mips big endian and mipsel little endian? Thefore `mips | mips-musl` would be better.

Comment (pbui) [reply common/build-style/meson.sh]: I've changed this to `mips|mips-musl`

File: common/build-style/meson.sh (hunk 2 above)
Comment (pbui) [reply common/build-style/meson.sh]: I've renamed this to `xbps_meson.cross`.

File: common/build-style/meson.sh (hunk 4 above)
Comment (pbui) [reply common/build-style/meson.sh]: Yes, I've switched this to `${CC}`, `${CXX}`, etc. and tested it on the ncmpc package with armv7l and arm7l-musl.

**Comment by pbui** [comment]
I've made the changes suggested by @Vaelatern and @Johnnynator.  

I've pushed a squashed commit... I'm not sure I was suppose to do that since it collapsed the review above.

**Comment by Gottox** [comment]
thanks, I'll give it a try with the latest Gnome update (https://github.com/voidlinux/void-packages/issues/7713).

File: common/build-style/meson.sh [hunk 5]
```diff
@@ -4,6 +4,54 @@
...
+cpu = '${_MESON_TARGET_CPU}'
+endian = '${_MESON_TARGET_ENDIAN}'
+EOF
+		configure_args+=" --cross-file=${meson_crossfile}"
+
+		# Meson tries to compile natively with CC, CXX, so when cross
+		# compiling, we need to set those to the host versions.
+		export CC=${CC_host} CXX=${CXX_host}
+
+		unset _MESON_TARGET_CPU _MESON_TARGET_ENDIAN
```
Comment (Gottox) [inline common/build-style/meson.sh]: declare those vars with `local _MESON_TARGET_CPU= _MESON_TARGET_ENDIAN=` at the top.

**Comment by Gottox** [comment]
I'd like to merge. Objections?

**Pull request merged**